File: livingwell-frontend/src/types/calculations.ts
export interface NOIInput {
  gross_potential_revenue: number;
  vacancy_rate?: number;
  operating_expenses?: number;
  property_tax?: number;
  insurance?: number;
  management_fee_rate?: number;
  replacement_reserves?: number;
}

export interface NOIResult {
  gross_potential_revenue: number;
  vacancy_loss: number;
  vacancy_rate: number;
  effective_gross_income: number;
  operating_expenses: number;
  property_tax: number;
  insurance: number;
  management_fee: number;
  management_fee_rate: number;
  replacement_reserves: number;
  total_expenses: number;
  noi: number;
}

export interface DSCRResult {
  noi: number;
  annual_debt_service: number;
  dscr: number | null;
  health: string;
  message: string;
}

export interface LTVResult {
  outstanding_debt: number;
  property_value: number;
  ltv_percent: number | null;
  equity_percent: number | null;
  equity_value: number | null;
  risk: string;
  message: string;
}

export interface IRRResult {
  irr_decimal: number | null;
  irr_percent: number | null;
  cash_flows: number[];
  message: string;
}

export interface PropertyFinancialSummary {
  property_id: number;
  property_name: string;
  noi: NOIResult | null;
  dscr: DSCRResult | null;
  ltv: LTVResult | null;
  cap_rate_percent: number | null;
  cash_on_cash_percent: number | null;
  total_debt_outstanding: number;
  total_equity: number;
  annual_debt_service: number;
}

File: backend/app/schemas/calculations.py
from pydantic import BaseModel
from typing import Optional


class NOIInput(BaseModel):
    gross_potential_revenue: float
    vacancy_rate: float = 0.05
    operating_expenses: float = 0.0
    property_tax: float = 0.0
    insurance: float = 0.0
    management_fee_rate: float = 0.04
    replacement_reserves: float = 0.0


class NOIResult(BaseModel):
    gross_potential_revenue: float
    vacancy_loss: float
    vacancy_rate: float
    effective_gross_income: float
    operating_expenses: float
    property_tax: float
    insurance: float
    management_fee: float
    management_fee_rate: float
    replacement_reserves: float
    total_expenses: float
    noi: float


class DSCRInput(BaseModel):
    noi: float
    annual_debt_service: float


class DSCRResult(BaseModel):
    noi: float
    annual_debt_service: float
    dscr: Optional[float]
    health: str
    message: str


class LTVInput(BaseModel):
    outstanding_debt: float
    property_value: float


class LTVResult(BaseModel):
    outstanding_debt: float
    property_value: float
    ltv_percent: Optional[float]
    equity_percent: Optional[float] = None
    equity_value: Optional[float] = None
    risk: str
    message: str


class IRRInput(BaseModel):
    cash_flows: list[float]


class IRRResult(BaseModel):
    irr_decimal: Optional[float]
    irr_percent: Optional[float]
    cash_flows: list[float]
    message: str


class PropertyFinancialSummary(BaseModel):
    property_id: int
    property_name: str
    noi: Optional[NOIResult] = None
    dscr: Optional[DSCRResult] = None
    ltv: Optional[LTVResult] = None
    cap_rate_percent: Optional[float] = None
    cash_on_cash_percent: Optional[float] = None
    total_debt_outstanding: float = 0.0
    total_equity: float = 0.0
    annual_debt_service: float = 0.0
